packages/dds/tree/src/schema-stored/schema.ts

```diff
@@ -249,17 +208,32 @@ export interface Named<TName> {
 
 export type NamedTreeSchema = TreeSchema & Named<TreeSchemaIdentifier>;
 
-export interface SchemaRepository {
+/**
+ * View of schema data that can be stored in a document.
+ *
+ * Note: the owner of this may modify it over time:
+ * thus if needing to hand onto a specific version, make a copy.
+ */
+export interface SchemaDataReader {
+    readonly globalFieldSchema: ReadonlyMap<GlobalFieldKey, FieldSchema>;
+    readonly treeSchema: ReadonlyMap<TreeSchemaIdentifier, TreeSchema>;
+}
+
+/**
+ * Policy from the app for interpreting the stored schema.
+ * The app must ensure consistency for all users of the document.
+ */
+export interface SchemaPolicy {
     /**
-     * All fields with the specified GlobalFieldKey must comply with the returned schema.
+     * Schema used when there is no schema explicitly specified for an identifier.
+     * Typically a "never" schema which forbids any nodes with that type.
      */
-    lookupGlobalFieldSchema(key: GlobalFieldKey): FieldSchema;
+    readonly defaultTreeSchema: TreeSchema;
 
     /**
-     * All trees with the specified identifier must comply with the returned schema.
+     * Schema used when there is no schema explicitly specified for an identifier.
+     * Typically an "empty" schema which forbids any field with that type from having children.
+     * TODO: maybe this must be an empty field? Anything else might break things.
      */
-    lookupTreeSchema(identifier: TreeSchemaIdentifier): TreeSchema;
-
-    readonly globalFieldSchema: ReadonlyMap<GlobalFieldKey, FieldSchema>;
-    readonly treeSchema: ReadonlyMap<TreeSchemaIdentifier, TreeSchema>;
+    readonly defaultGlobalFieldSchema: FieldSchema;
```

Comment:
I'm a bit confused on why we have a separate `defaultGlobalFieldSchema` as part of `SchemaPolicy`. Is this just because the `FieldSchema`s stored in `TreeSchema` are mapped against specific field names and we want this to apply to all fields with some field kind? If so, it seems insufficient to have just one `FieldSchema`. I also don't see the value that `SchemaPolicy` adds over just `TreeSchema` with an additional field.

Reply:
This policy is for an entire application, not for a specific TreeSchema, and applies to every globalFieldIdentifier that could ever exist that does not currently have a stored schema.

Effectively an empty document starts with all possible tree and field schema existing, but defaulting to never and empty respectively.  This means that we can just support schema upgrades and don't need to support schema creation/insertions. This greatly reduces the amount of schema edit types we need to support, and simplifies the APIs, but does require these defaults to exist somewhere, and that can't be in  schema-stored since they depend on specific field kinds, so it has to be injected via this policy object.


I don't understand what you are suggesting as an alternative, but I don't think it addresses the same thing.

Reply:
Oh, I'm not really suggesting an alternative. I just don't quite understand how everything fits together in this schema system just yet, especially in the concrete implementation. I still don't really understand but I'll have more targeted questions for you after I go through everything some more 😄.